### PR TITLE
Use Vuex for new History

### PR DESCRIPTION
**Pull request by guerler**
This PR transitions from UrlProviders to Vuex Providers. We are adding stores for history items, collection elements and datasets. The stores are attached to a generic `store provider` which is imported into Vue components and passes custom props to the store. Changes to the props trigger a `store action` and a `store getter` provides the updated results. The stores trigger individual requests to obtain the data. Additionally a continuous request (every 3 seconds) is performed by the `history changed items` store watching the current history. The changed items store passes the list of changed items to the `history items`, `dataset` and `collection elements` stores by executing actions. All of these stores cache all responses and as such also need to apply client side filtering. Overall, this model follows suggestions described in the second strategy of #13510.

image

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by guerler** [comment]
@galaxybot test this

**Comment by bgruening** [comment]
I tested this branch and its works very well for me. Tested under Chrome and FF, I do not see a lot of requests and the scrolling is for me smooth. I see rendering glitches, but this is hard to describe and can be addressed later imho. With this PR I think its definitely on par with the old history, not already better.

File: client/src/components/Form/Elements/FormNumber.vue [hunk 1]
```diff
@@ -34,12 +34,12 @@ export default {
             validator: (prop) => ["integer", "float"].includes(prop.toLowerCase()),
         },
         min: {
-            type: Number,
+            type: [Number, String],
```
Comment (mvdbeek) [inline client/src/components/Form/Elements/FormNumber.vue:37]: Why did you do this ? That smells like we're hiding a problem here.

**Review by mvdbeek — APPROVED** [review]
Congratulations @guerler, this is working extremely well. I am a little worried about performance problems on the server side if we are getting the detailed view without filters for updates. But it also looks like we can move to a batch update API pretty easily with this setup, so let's get this in and on a real server.

I would also like to see some store tests as well. They're probably a pain, but I also think they'll be useful.

File: client/src/components/providers/StoreProvider.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { mapActions, mapGetters } from "vuex";
```
[the remark this answers is not in the document]
Comment (mvdbeek) [reply client/src/components/providers/StoreProvider.js]: Never mind that comment, was from an older review ...

**Comment by mvdbeek** [comment]
Actually it doesn't look like query is quite right. If you change the name of a history item in another tab this never comes in from the API. That does work in the old history and I think that is very important.

File: client/src/store/historyStore/historyChangedItemsStore.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import { urlData } from "utils/url";
+
+let fetching = false;
+let lastDate = new Date();
+const limit = 100;
+const throttlePeriod = 3000;
+
+const actions = {
+    fetchHistoryChangedItems: async ({ commit, dispatch }, { historyId }) => {
+        const params = `limit=${limit}&q=update_time-ge&qv=${lastDate.toISOString()}`;
```
Comment (mvdbeek) [inline client/src/store/historyStore/historyChangedItemsStore.js]:
```suggestion
        const params = `limit=${limit}&q=update_time-ge&qv=${encodeURIComponent(lastDate.toISOString())}`;
```
that fixes the update time issue for me. However the API now keeps returning the updated dataset, even though the update time increments.

Comment (mvdbeek) [reply client/src/store/historyStore/historyChangedItemsStore.js]: Nope, not sure what I saw there, with the suggestion it seems to work.

File: client/src/components/Form/Elements/FormNumber.vue (hunk 1 above)
Comment (guerler) [reply client/src/components/Form/Elements/FormNumber.vue:37]: Yes, correct. The form dictionaries from the visualization serialization in the backend provides these values as strings not as numeric values, leading to console prop mismatch errors. This fix resolves this issue but the number component has to be revisited to properly show the alert for values which are not in the min/max interval.

File: client/src/store/historyStore/historyChangedItemsStore.js (hunk 3 above)
Comment (guerler) [reply client/src/store/historyStore/historyChangedItemsStore.js]: I tried it and it works for me too. Thx for the fix.

**Comment by guerler** [comment]
Thank you for the review @mvdbeek, your descriptions in #13510 was very helpful. I am glad that the model is going into the right direction. I agree that we do need much more tests for the new history, including for these stores. I definitely intend to add tests throughout for the new history in follow-ups before the next release. Thx again for the fix, I tested changes across tabs and that worked well.

File: client/src/store/historyStore/historyChangedItemsStore.js (hunk 3 above)
Comment (dannon) [reply client/src/store/historyStore/historyChangedItemsStore.js]: I think @mvdbeek is right here; the colons in the iso date string should be encoded -- it doesn't look like this got applied?

**Review by dannon — APPROVED** [review]
This looks great, and is working well for me.  I think we can probably drop the dataset fetch debugging now (you can see it in network tab anyway), but I'm going to go ahead and merge this after tests finish as a sanity check (two failed, rerunning now) after re-adding @mvdbeek's suggestion.

**Pull request merged**